### PR TITLE
R tensorflow example

### PR DESCRIPTION
**Pull request by jnolis**
This is an example of using a TensorFlow model with a GPU in R.

it uses the new `saturn-rstudio-tensorflow` GPU image, which won't exist until [this PR](https://github.com/saturncloud/images/pull/294) gets through the system

### CODE REVIEW

**Comment by jnolis** [comment]
(cc @forana )

File: examples/r-tensorflow/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Train an TensorFlow model with R on a GPU
+
+This example uses the reticulate, keras, and tensorflow packages of R to train a model on a GPU.
+To run it, use the `saturn-rstudio-tensorflow` image. If you would rather train it on a GPU you
```
Comment (nathanballou) [inline examples/r-tensorflow/README.md]:
```suggestion
To run it, use the `saturn-rstudio-tensorflow` image. If you would rather train it on a CPU you
```

File: examples/r-tensorflow/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Train an TensorFlow model with R on a GPU
+
+This example uses the reticulate, keras, and tensorflow packages of R to train a model on a GPU.
```
Comment (nathanballou) [inline examples/r-tensorflow/README.md]:
```suggestion
This example uses the reticulate, Keras, and TensorFlow packages of R to train a model on a GPU.
```

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+R has the capability to train models with TensorFlow and Keras using packages
+developed by RStudio. On the backend, these packages are running TensorFlow in
+Python, and then the reticulate package converts the Python objects to R. As an
+R user, that is largely abstracted away so you can primarily think in term of
+the R language you're used to.
+
+While often getting Python, R, TensorFlow, and the GPU drivers to all be the
+correct versions and work together, Saturn Cloud provides a convenient
+`saturn-rstudio-tensorflow`image that is preconfigured to train the models on a
+GPU. If you'd rather train a model on a cpu, you can use the `saturn-rstudio`
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]:
```suggestion
GPU. If you'd rather train a model on a CPU, you can use the `saturn-rstudio`
```

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+image and install the keras R package yourself.
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]: this is different in the readme. Which one is accurate?

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+```{r data_matrix}
+text_matrix <-
+  subsequence_data %>%
+  map(~ character_lookup$character_id[match(.x,character_lookup$character)]) %>% 
+  pad_sequences(maxlen = max_length+1) %>% 
+  to_categorical(num_classes = num_characters) 
+```
+
+Once that's complete we split the data into the X 3D-matrix of model input and
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]:
```suggestion
Once that's complete, we split the data into the 3D-matrix of model input (X) and
```

Maybe a bit clearer? Up to you.

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+y matrix of targets. This last splitting is done on an nparray,
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]:
```suggestion
matrix of targets (y). This last splitting is done on an nparray,
```

"

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+x_name <- np$delete(text_matrix, as.integer(max_length), 1L) 
+y_name <- np$delete(text_matrix, as.integer((1:max_length)-1), 1L)$squeeze()
+```
+
+### Create the model
+
+Next we define the layers of the Keras model. This model has 2 LSTM layers to
+find the patterns in the names, a dense layer to predict a value for each
+possible next character, and a softmax activation to turn
+those values into probabilities. Since this is a multiclassification problem,
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]:
```suggestion
those values into probabilities. Since this is a multiclass classification problem,
```

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+the loss is categorical crossentropy.
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]:
```suggestion
the loss is categorical cross-entropy.
```

File: examples/r-tensorflow/r-tensorflow.Rmd [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+model <- keras_model(inputs = input, outputs = output) %>% 
+  compile(
+    loss = 'categorical_crossentropy',
+    optimizer = "adam"
+  )
+```
+
+### Train the model
+
+Once the model is defined we can then train it.
```
Comment (nathanballou) [inline examples/r-tensorflow/r-tensorflow.Rmd]:
```suggestion
Once the model is defined, we can train it.
```

**Review by nathanballou — APPROVED** [review]
Just some grammatical feedback.
